src/posix/platform/udp.cpp

```diff
@@ -116,7 +116,7 @@ static otError transmitPacket(int aFd, uint8_t *aPayload, uint16_t aLength, cons
     cmsg = CMSG_FIRSTHDR(&msg);
 
     {
-        int hopLimit = (aMessageInfo.mHopLimit ? aMessageInfo.mHopLimit : -1);
+        int hopLimit = (aMessageInfo.mHopLimit ? aMessageInfo.mHopLimit : OPENTHREAD_CONFIG_IP6_HOP_LIMIT_DEFAULT);
```

Comment:
Is it better to take the destination scope into consideration? Maybe we should set this value only when the destination scope is larger than link local scope?

Reply:
It's not clear what value to set. For example, RFC 4861 uses IPv6 Hop Limit of 255 for link-local messages. Also, routers should never forward link-local messages regardless of IPv6 Hop Limit value.

The alternative is we explicitly set the IPv6 Hop Limit value whenever sending an IPv6 multicast message, but that seemed like a bigger change and propagates to CLI commands (e.g. ping).

In any case, I'd prefer to make the behavior consistent regardless of whether `PLATFORM_UDP` is enabled.

Thoughts?

Reply:
I think the kernel will correctly choose the hoplimit when we give it `-1` here for non-realm-local scopes. Can this issue be fixed by only setting OpenThread's default hoplimit for realm-local multicast?

Reply:
Are you sure? This is what I see on ubuntu 20.04.

```
00:54:25.800380 IP6 (flowlabel 0x5bb1c, hlim 1, next-header ICMPv6 (58) payload length: 64) fe80::42:acff:fe11:2 > ff05::1: [icmp6 sum ok] ICMP6, echo request, seq 6
00:54:40.613507 IP6 (flowlabel 0xbcade, hlim 1, next-header ICMPv6 (58) payload length: 64) fe80::42:acff:fe11:2 > ff08::1: [icmp6 sum ok] ICMP6, echo request, seq 1
```

It seems the IPv6 Hop Limit is always 1 regardless of scope.

Reply:
Also see: https://docs.oracle.com/cd/E19253-01/817-4415/sockets-13/index.html

> By default, IP multicast datagrams are sent with a hop limit of one, which prevents the datagrams from being forwarded beyond a single subnetwork.

Reply:
I just realized this is mainly used by OpenThread's internal services, and it will not affect application layer behavior on Linux. Thank you for sharing the findings!